src/rust/cryptography-x509-validation/src/lib.rs

```diff
@@ -11,6 +11,289 @@ pub mod policy;
+                    ))
+                }
+            }
+            (GeneralName::IPAddress(pattern), GeneralName::IPAddress(name)) => {
+                if let Some(pattern) = IPConstraint::from_bytes(pattern) {
+                    let name = IPAddress::from_bytes(name).unwrap();
+                    Ok(Applied(pattern.matches(&name)))
+                } else {
+                    Err(ValidationError::Other(
+                        "malformed IP name constraint".to_string(),
+                    ))
+                }
+            }
+            _ => Ok(Skipped),
+        }
+    }
+
+    fn apply_name_constraints(
+        &self,
+        constraints: &Vec<NameConstraints<'chain>>,
+        extensions: &Extensions<'chain>,
+    ) -> Result<(), ValidationError> {
+        if let Some(sans) = extensions.get_extension(&SUBJECT_ALTERNATIVE_NAME_OID) {
+            let sans: SubjectAlternativeName<'_> = sans.value()?;
+            for san in sans.clone() {
+                // If there are no applicable constraints, the SAN is considered valid so the default is true.
+                let mut permit = true;
+                for nc in constraints {
+                    if let Some(permitted_subtrees) = &nc.permitted_subtrees {
+                        for p in permitted_subtrees.unwrap_read().clone() {
+                            let status = self.apply_name_constraint(&p.base, &san)?;
+                            if status.is_applied() {
+                                permit = status.is_match();
+                                if permit {
+                                    break;
+                                }
+                            }
+                        }
+                    }
+                }
+                if !permit {
+                    return Err(ValidationError::Other(
+                        "no permitted name constraints matched SAN".into(),
+                    ));
+                }
+                for nc in constraints {
+                    if let Some(excluded_subtrees) = &nc.excluded_subtrees {
+                        for e in excluded_subtrees.unwrap_read().clone() {
+                            let status = self.apply_name_constraint(&e.base, &san)?;
+                            if status.is_match() {
+                                return Err(ValidationError::Other(
+                                    "excluded name constraint matched SAN".into(),
+                                ));
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn build_chain_inner(
+        &self,
+        working_cert: &'a Certificate<'chain>,
+        current_depth: u8,
+        is_leaf: bool,
+        extensions: &'a Extensions<'chain>,
+    ) -> Result<IntermediateChain<'chain>, ValidationError> {
+        if current_depth > self.policy.max_chain_depth {
```

Comment:
This depth makes me wonder: Does limbo include a test case that has two intermediates signing each other for an infinite loop?

Reply:
I'll check, but I don't believe so!

Reply:
Opened https://github.com/trailofbits/x509-limbo/pull/92 for this, and confirmed that our current implementation rejects this as expected.